{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import re\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import tarfile\n",
    "import os\n",
    "import sys\n",
    "import shutil\n",
    "import numpy as np\n",
    "import tarfile\n",
    "import io\n",
    "from io import StringIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class gdc_data:\n",
    "    \n",
    "    def __init__(self, c_type):\n",
    "        #Initialize the type of cancer for the database query\n",
    "        self.c_type = c_type\n",
    "        #initialize gene epression data matrix\n",
    "        self.data = pd.DataFrame()\n",
    "        #perform query and store data query\n",
    "        \n",
    "        #Initialize empty manifest data matrix\n",
    "        self.manifest = pd.DataFrame()\n",
    "\n",
    "    def gdc_query(self, cancer):\n",
    "        '''\n",
    "        Performs a query of the NCI genomic portal given a type of cancer.\n",
    "        Hepatocellular Carcinoma - LIHC\n",
    "        Returns the name of compressed tar.gz file, and a binary data file\n",
    "\n",
    "        '''\n",
    "        files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "        #Filters for the query, recieving all RNA-Seq, HTSeq-Count files for a specific cancer\n",
    "        filters = {\n",
    "            \"op\": \"and\",\n",
    "            \"content\":[\n",
    "                {\n",
    "                \"op\": \"in\",\n",
    "                \"content\":{\n",
    "                    \"field\": \"cases.project.project_id\",\n",
    "                    \"value\": [\"TCGA-\"+cancer]\n",
    "                    }\n",
    "                },\n",
    "                {\n",
    "                \"op\": \"in\",\n",
    "                \"content\":{\n",
    "                    \"field\": \"files.experimental_strategy\",\n",
    "                    \"value\": [\"RNA-Seq\"]\n",
    "                    }\n",
    "                },\n",
    "                {\n",
    "                \"op\": \"in\",\n",
    "                \"content\":{\n",
    "                    \"field\": \"files.analysis.workflow_type\",\n",
    "                    \"value\": [\"HTSeq - Counts\"]\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "\n",
    "        # Here a GET is used, so the filter parameters should be passed as a JSON string.\n",
    "        params = {\n",
    "            \"filters\": json.dumps(filters),\n",
    "            \"fields\": \"file_id\",\n",
    "            \"format\": \"JSON\",\n",
    "            \"size\": \"10\" #Set to the first 10 files for developing\n",
    "            }\n",
    "\n",
    "        response = requests.get(files_endpt, params = params)\n",
    "        file_uuid_list = []\n",
    "\n",
    "        # This step populates the download list with the file_ids from the previous query\n",
    "        for file_entry in json.loads(response.content.decode(\"utf-8\"))[\"data\"][\"hits\"]:\n",
    "            file_uuid_list.append(file_entry[\"file_id\"])\n",
    "\n",
    "        data_endpt = \"https://api.gdc.cancer.gov/data\"\n",
    "\n",
    "        params = {\"ids\": file_uuid_list}\n",
    "\n",
    "        response = requests.post(data_endpt, data = json.dumps(params), headers = {\"Content-Type\": \"application/json\"})\n",
    "\n",
    "        response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "\n",
    "        file_name = re.findall(\"filename=(.+)\", response_head_cd)[0]\n",
    "\n",
    "        #with open(file_name, \"wb\") as output_file:\n",
    "         #   output_file.write(response.content)\n",
    "\n",
    "        return file_name,response\n",
    "    \n",
    "    def data_write(self):\n",
    "        #Acquires the file_name and the data from the query in \"response\"\n",
    "        self.file_name,self.response = self.gdc_query(self.c_type)\n",
    "        \n",
    "        #location of compressed data file\n",
    "        comp_data = os.path.join(\"./test_data/\",file_name)\n",
    "        \n",
    "        #Opens a file named after the file_name, and writes the contents of the query to the file\n",
    "        if not os.path.exists(\"./test_data/\"):\n",
    "            os.makedirs(\"./test_data/\")\n",
    "        with open(comp_data, \"wb\") as output_file:\n",
    "            output_file.write(response.content)\n",
    "        \n",
    "        #Unzips the tar.gz file that was generated into a test data subdirectory\n",
    "        with tarfile.open(comp_data) as tar:\n",
    "            tar.extractall(\"./test_data/\")\n",
    "            tar.close()\n",
    "        \n",
    "        #Stores the manifest of the data \n",
    "        manifest = pd.read_table(\"./test_data/MANIFEST.txt\",sep=\"\\t\")\n",
    "        \n",
    "        uncomp_data = os.path.join(os.getcwd(),\"./test_data/uncompressed_data/\")\n",
    "        \n",
    "        if not os.path.exists(uncomp_data):\n",
    "            os.makedirs((\"./test_data/uncompressed_data\"))\n",
    "            \n",
    "        for subdir, dirs, files in os.walk(\"./test_data/\"):\n",
    "            for file in files:\n",
    "                if file[-4:] == \"s.gz\":\n",
    "                    with gzip.open(os.path.join(subdir,file),'rb') as f:\n",
    "                        file_content = f.read().decode(\"utf-8\")\n",
    "                        df = pd.read_csv(StringIO(file_content),sep=\"\\t\",header=None).set_index(0)\n",
    "                        df.columns = [files[0]]\n",
    "                        data = pd.concat([data,df],axis=1)\n",
    "                        #df.to_csv(os.path.join(\"./test_data/uncompressed_data/\",file[:-3]),header=False,sep=\",\",index=False)\n",
    "        \n",
    "    def data_save(self):\n",
    "        #initialize/clear gene epression data matrix\n",
    "        self.data = pd.DataFrame()\n",
    "        #Stores the manifest of the data \n",
    "        self.manifest = pd.read_table(\"./test_data/MANIFEST.txt\",sep=\"\\t\")\n",
    "            \n",
    "        for subdir, dirs, files in os.walk(\"./test_data/\"):\n",
    "            for file in files:\n",
    "                if file[-4:] == \"s.gz\":\n",
    "                    with gzip.open(os.path.join(subdir,file),'rb') as f:\n",
    "                        file_content = f.read().decode(\"utf-8\")\n",
    "                        df = pd.read_csv(StringIO(file_content),sep=\"\\t\",header=None).set_index(0)\n",
    "                        df.columns = [files[0]]\n",
    "                        self.data = pd.concat([self.data,df],axis=1)\n",
    "                                \n",
    "                    \n",
    "    def data_add(self,data):\n",
    "            self.data = data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test = gdc_data(\"LIHC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "test.data_save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7807318b-32d2-480e-805a-911d90f43e52.htseq.counts.gz</th>\n",
       "      <th>f6ae6ac1-3e00-4021-a6e7-fbb0d5f12836.htseq.counts.gz</th>\n",
       "      <th>554f6de3-63c7-47b1-a75a-dcfc73f54e96.htseq.counts.gz</th>\n",
       "      <th>687e7d1d-99eb-4bf9-9fa3-49e324ef32c3.htseq.counts.gz</th>\n",
       "      <th>e8aae9c3-1e7d-4f47-bd78-c96f55f992db.htseq.counts.gz</th>\n",
       "      <th>a2a12535-77ea-414b-bd36-d9328e019d05.htseq.counts.gz</th>\n",
       "      <th>f32c1def-c5c6-4076-966e-ae5f7233060a.htseq.counts.gz</th>\n",
       "      <th>7097ca31-8322-44a6-8fe7-f0c67bbf88fe.htseq.counts.gz</th>\n",
       "      <th>5fe28ffa-63af-4a8d-8512-b0742b4cded4.htseq.counts.gz</th>\n",
       "      <th>0fc6f38a-62da-4c2f-8a72-5c34b77656e5.htseq.counts.gz</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <td>2194</td>\n",
       "      <td>2893</td>\n",
       "      <td>3600</td>\n",
       "      <td>3920</td>\n",
       "      <td>16166</td>\n",
       "      <td>3093</td>\n",
       "      <td>8685</td>\n",
       "      <td>1976</td>\n",
       "      <td>4740</td>\n",
       "      <td>4455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <td>1313</td>\n",
       "      <td>1048</td>\n",
       "      <td>2769</td>\n",
       "      <td>2116</td>\n",
       "      <td>1151</td>\n",
       "      <td>1425</td>\n",
       "      <td>2054</td>\n",
       "      <td>502</td>\n",
       "      <td>3645</td>\n",
       "      <td>955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <td>432</td>\n",
       "      <td>936</td>\n",
       "      <td>442</td>\n",
       "      <td>854</td>\n",
       "      <td>912</td>\n",
       "      <td>441</td>\n",
       "      <td>779</td>\n",
       "      <td>431</td>\n",
       "      <td>492</td>\n",
       "      <td>432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <td>153</td>\n",
       "      <td>392</td>\n",
       "      <td>237</td>\n",
       "      <td>2657</td>\n",
       "      <td>1669</td>\n",
       "      <td>76</td>\n",
       "      <td>191</td>\n",
       "      <td>134</td>\n",
       "      <td>611</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    7807318b-32d2-480e-805a-911d90f43e52.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               2194      \n",
       "ENSG00000000005.5                                                   0      \n",
       "ENSG00000000419.11                                               1313      \n",
       "ENSG00000000457.12                                                432      \n",
       "ENSG00000000460.15                                                153      \n",
       "\n",
       "                    f6ae6ac1-3e00-4021-a6e7-fbb0d5f12836.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               2893      \n",
       "ENSG00000000005.5                                                   1      \n",
       "ENSG00000000419.11                                               1048      \n",
       "ENSG00000000457.12                                                936      \n",
       "ENSG00000000460.15                                                392      \n",
       "\n",
       "                    554f6de3-63c7-47b1-a75a-dcfc73f54e96.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               3600      \n",
       "ENSG00000000005.5                                                   0      \n",
       "ENSG00000000419.11                                               2769      \n",
       "ENSG00000000457.12                                                442      \n",
       "ENSG00000000460.15                                                237      \n",
       "\n",
       "                    687e7d1d-99eb-4bf9-9fa3-49e324ef32c3.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               3920      \n",
       "ENSG00000000005.5                                                  13      \n",
       "ENSG00000000419.11                                               2116      \n",
       "ENSG00000000457.12                                                854      \n",
       "ENSG00000000460.15                                               2657      \n",
       "\n",
       "                    e8aae9c3-1e7d-4f47-bd78-c96f55f992db.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                              16166      \n",
       "ENSG00000000005.5                                                   1      \n",
       "ENSG00000000419.11                                               1151      \n",
       "ENSG00000000457.12                                                912      \n",
       "ENSG00000000460.15                                               1669      \n",
       "\n",
       "                    a2a12535-77ea-414b-bd36-d9328e019d05.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               3093      \n",
       "ENSG00000000005.5                                                   0      \n",
       "ENSG00000000419.11                                               1425      \n",
       "ENSG00000000457.12                                                441      \n",
       "ENSG00000000460.15                                                 76      \n",
       "\n",
       "                    f32c1def-c5c6-4076-966e-ae5f7233060a.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               8685      \n",
       "ENSG00000000005.5                                                   1      \n",
       "ENSG00000000419.11                                               2054      \n",
       "ENSG00000000457.12                                                779      \n",
       "ENSG00000000460.15                                                191      \n",
       "\n",
       "                    7097ca31-8322-44a6-8fe7-f0c67bbf88fe.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               1976      \n",
       "ENSG00000000005.5                                                   2      \n",
       "ENSG00000000419.11                                                502      \n",
       "ENSG00000000457.12                                                431      \n",
       "ENSG00000000460.15                                                134      \n",
       "\n",
       "                    5fe28ffa-63af-4a8d-8512-b0742b4cded4.htseq.counts.gz  \\\n",
       "0                                                                          \n",
       "ENSG00000000003.13                                               4740      \n",
       "ENSG00000000005.5                                                   0      \n",
       "ENSG00000000419.11                                               3645      \n",
       "ENSG00000000457.12                                                492      \n",
       "ENSG00000000460.15                                                611      \n",
       "\n",
       "                    0fc6f38a-62da-4c2f-8a72-5c34b77656e5.htseq.counts.gz  \n",
       "0                                                                         \n",
       "ENSG00000000003.13                                               4455     \n",
       "ENSG00000000005.5                                                   1     \n",
       "ENSG00000000419.11                                                955     \n",
       "ENSG00000000457.12                                                432     \n",
       "ENSG00000000460.15                                                213     "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
